pvanalytics/features/clipping.py

```diff
@@ -88,3 +88,115 @@ def levels(ac_power, window=4, fraction_in_window=0.75,
         flags = flags | _label_clipping(temp, window=window,
                                         frac=fraction_in_window)
     return flags
+
+
+def _daytime_powercurve(ac_power):
+    # return the 99.5% quantile of daytime power data after removing
+    # night time data.
+    #
+    # Copyright (c) 2020 Alliance for Sustainable Energy, LLC.
+    power = ac_power.to_frame()
+    power_column = power.columns[0]
+    power['minutes'] = power.index.hour * 60 + power.index.minute
+
+    power['positive_power'] = ac_power
+    power.loc[power.positive_power >= 0, 'positive_power'] = 1
+    powerfreq = power.groupby('minutes').positive_power.sum()
+    daytimes = powerfreq[powerfreq >= powerfreq.quantile(0.25)].index
+    daytime_power = power[power.minutes.isin(daytimes)]
+
+    return daytime_power.groupby('minutes')[power_column].quantile(0.995)
+
+
+def _clipped(power, derivative, clip_derivative, minimum):
+    # test whether a `power` is greater than `minimum` and
+    # `derivative` is less than `clip_derivative`
+    return (np.abs(derivative) <= clip_derivative) and (power > minimum)
+
+
+def _clipping_power(ac_power, clip_derivative=0.0035, freq=None):
+    # Returns the clipping power threshold or None if no clipping is
+    # identified in the data.
+    #
+    # Copyright (c) 2020 Alliance for Sustainable Energy, LLC.
+    if not freq:
+        freq = pd.Timedelta(pd.infer_freq(ac_power.index)).seconds * 60
+    elif isinstance(freq, str):
+        freq = pd.Timedelta(freq).seconds * 60
+
+    # Use the derivative of the 99.5% quantile of daytime power at
+    # each minute to identify clipping.
+    powercurve = _daytime_powercurve(ac_power)
+    normalized_power = powercurve / powercurve.max()
+    power_derivative = (normalized_power.diff()
+                        / normalized_power.index.to_series().diff()) * freq
+    power_median = powercurve.median()
+
+    powersum = 0
+    count = 0
+    longest_powersum = 0
+    longest_count = 0
+    for derivative, power in zip(power_derivative, powercurve):
```

Comment:
seems like a use for `Series.rolling` somehow.

Reply:
I couldn't see a way to do it with Series.rolling because the count resets under certain conditions. It is kind of a weird combination of `Series.expanding` and `Series.rolling` I will dig in a little deeper to both and see if there is a way to get rid of the loop.

Reply:
I have a nice solution to this that uses cumsum and some other tricks to make the code more readable. Probably doesn't have much of a performance impact since the loop was executing at most 1440 iterations (# of minutes in the day).